File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.cpp
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> result;
        if (words.empty() || s.empty()) return result;
        
        int wordLen = words[0].length();
        int wordCount = words.size();
        int totalLen = wordLen * wordCount;
        if (s.length() < totalLen) return result;
        
        unordered_map<string, int> wordFreq;
        for (const string& w : words) {
            wordFreq[w]++;
        }
        
        // Slide over each possible starting offset modulo wordLen
        for (int offset = 0; offset < wordLen; ++offset) {
            unordered_map<string, int> seen;
            int left = offset;
            int matched = 0;
            
            for (int right = offset; right + wordLen <= s.length(); right += wordLen) {
                string curr = s.substr(right, wordLen);
                
                if (wordFreq.count(curr)) {
                    seen[curr]++;
                    if (seen[curr] == wordFreq[curr]) {
                        matched++;
                    }
                    
                    // Shrink window if we have more than needed
                    while (seen[curr] > wordFreq[curr]) {
                        string leftWord = s.substr(left, wordLen);
                        if (seen[leftWord] == wordFreq[leftWord]) {
                            matched--;
                        }
                        seen[leftWord]--;
                        left += wordLen;
                    }
                    
                    // If we have matched all words, record starting index
                    if (matched == wordFreq.size()) {
                        result.push_back(left);
                        // Move left one word to continue searching
                        string leftWord = s.substr(left, wordLen);
                        if (seen[leftWord] == wordFreq[leftWord]) {
                            matched--;
                        }
                        seen[leftWord]--;
                        left += wordLen;
                    }
                } else {
                    // Reset window because current word is not in words
                    seen.clear();
                    matched = 0;
                    left = right + wordLen;
                }
            }
        }
        return result;
    }
};

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.java
import java.util.*;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s == null || s.length() == 0 || words == null || words.length == 0) {
            return result;
        }
        
        int wordLen = words[0].length();
        int numWords = words.length;
        int totalLen = wordLen * numWords;
        int n = s.length();
        
        if (n < totalLen) return result;
        
        // Build required frequency map
        Map<String, Integer> required = new HashMap<>();
        for (String word : words) {
            required.put(word, required.getOrDefault(word, 0) + 1);
        }
        
        // Try all possible starting offsets (0 to wordLen - 1)
        for (int i = 0; i < wordLen; i++) {
            int left = i;
            int matched = 0;
            Map<String, Integer> seen = new HashMap<>();
            
            // Slide word by word
            for (int right = i; right + wordLen <= n; right += wordLen) {
                String word = s.substring(right, right + wordLen);
                
                if (required.containsKey(word)) {
                    seen.put(word, seen.getOrDefault(word, 0) + 1);
                    
                    if (seen.get(word) <= required.get(word)) {
                        matched++;
                    }
                    
                    // Shrink window if we have too many of this word
                    while (seen.get(word) > required.get(word)) {
                        String leftWord = s.substring(left, left + wordLen);
                        seen.put(leftWord, seen.get(leftWord) - 1);
                        if (seen.get(leftWord) < required.get(leftWord)) {
                            matched--;
                        }
                        left += wordLen;
                    }
                    
                    // Check if we found a valid window
                    if (matched == numWords) {
                        result.add(left);
                        
                        // Remove the leftmost word to look for the next valid window
                        String leftWord = s.substring(left, left + wordLen);
                        seen.put(leftWord, seen.get(leftWord) - 1);
                        if (seen.get(leftWord) < required.get(leftWord)) {
                            matched--;
                        }
                        left += wordLen;
                    }
                } else {
                    // Invalid word found, reset the window
                    seen.clear();
                    matched = 0;
                    left = right + wordLen;
                }
            }
        }
        
        return result;
    }
}